firebase-authexchange/src/main/kotlin/com/google/firebase/authexchange/network/Routes.kt

```diff
@@ -0,0 +1,105 @@
+package com.google.firebase.authexchange.network
+
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+/**
+ * Defines what an HTTP request may look like when interacting with the backend API for
+ * [FirebaseAuthExchange][com.google.firebase.authexchange.FirebaseAuthExchange].
+ */
+@Serializable internal sealed interface Request
+
+/** Defines what an HTTP response may look like when submitting a [Request]. */
+@Serializable internal sealed interface Response
+
+/**
+ * OIDC Credential.
+ *
+ * Useful for apps with external login already set-up, otherwise known as a headless OIDC flow.
+ *
+ * @property idToken JWT encoded OIDC token returned from a third party provider
+ */
+@Serializable internal data class ImplicitCredentialsP(val idToken: String)
+
+/**
+ * Auth Token that can be used to access certain Firebase Services.
+ *
+ * This is merely the proto type for the network layer, and is completely separate from the user
+ * facing [AuthExchangeToken][com.google.firebase.authexchange.AuthExchangeToken].
+ *
+ * @property accessToken signed JWT containing claims that identify a user
+ * @property timeToLive the duration from the time this token is minted until its expiration
+ */
+@Serializable
+internal data class AuthExchangeTokenP(
+  val accessToken: String,
+  @SerialName("ttl") val timeToLive: String
+)
+
+/**
+ * Request header for the `/ExchangeInstallationAuthToken` endpoint.
+ *
+ * @see ExchangeTokenResponseP
+ *
+ * @property token relative resource name of the audience project and location
+ * @property installationAuthToken the installation token issued to the app
+ */
+@Serializable
+internal data class ExchangeInstallationAuthTokenRequestP(
+  val token: String,
+  val installationAuthToken: String
+) : Request
+
+/**
+ * Request header for the `/ExchangeCustomToken` endpoint.
+ *
+ * @see ExchangeTokenResponseP
+ *
+ * @property token relative resource name of the audience project and location
+ * @property customToken a custom JWT token signed with the developer's credentials
+ */
+@Serializable
+internal data class ExchangeCustomTokenRequestP(val token: String, val customToken: String) :
+  Request
+
+/**
+ * Request header for the `/ExchangeOidcToken` endpoint.
+ *
+ * @see ExchangeOidcTokenResponseP
+ *
+ * @property token relative resource name of the audience project and location
+ * @property providerId the display name or id of the OIDC provider
+ * @property implicitCredentials JWT token from the OIDC provider, provided by the developer
+ */
+@Serializable
+internal data class ExchangeOidcTokenRequestP(
+  val token: String,
+  val providerId: String,
+  val implicitCredentials: ImplicitCredentialsP
```

Comment:
The `ExchangeOidcTokenRequest` can take either implicit credentials or auth code credentials (see the [API  proto](http://google3/google/firebase/authexchange/v1alpha/token_exchange.proto;l=142;rcl=500590766)). Is there a way to represent this as a `oneof`?

Reply:
For sure, although do we use it on our end? I initially had the following comment:
```kotlin
 * While this isn't the only OIDC flow that the API supports, it's the only flow that the mobile SDK
 * supports- so there is no base class for credentials.
 ```
Because I was under the impression we don't utilize `AuthCodeCredentials` on the SDK?

Reply:
The auth code OIDC flow is the "headful" flow (detailed [here](https://docs.google.com/document/d/1qpekuA5yyImI8MCAsNmt3HIMa6uHzlGuqmEtLgViAVI/edit?resourcekey=0-8g6zKHNr6Ii_RsxQ3j6GTw#)).